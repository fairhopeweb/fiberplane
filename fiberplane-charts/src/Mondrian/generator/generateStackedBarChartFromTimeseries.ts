import type {
  AbstractChart,
  Axis,
  TimeseriesSourceData,
  Shape,
  ShapeList,
} from "../types";
import {
  StackedChartBuckets,
  calculateBucketsAndAxesForStackedChart,
  calculateBarWidth,
  calculateSmallestTimeInterval,
  extendAxisWithInterval,
  getTimeFromTimestamp,
  normalizeAlongLinearAxis,
} from "./utils";
import type { Metric, Timeseries } from "../../providerTypes";
import { compact } from "../../utils";

export function generateStackedBarChartFromTimeseries(
  input: TimeseriesSourceData,
): AbstractChart<Timeseries, Metric> {
  const { buckets, isPercentage, xAxis, yAxis } =
    calculateBucketsAndAxesForStackedChart(input);

  const interval = calculateSmallestTimeInterval(buckets);
  if (interval) {
    extendAxisWithInterval(xAxis, interval);
  }

  const barWidth = calculateBarWidth(xAxis, interval ?? 0, 1);
  const barArgs: BarArgs = { barWidth, buckets, isPercentage, xAxis, yAxis };

  const shapeLists: Array<ShapeList<Timeseries, Metric>> =
    input.timeseriesData.map((timeseries) => ({
      shapes: timeseries.visible
        ? compact(
            timeseries.metrics.map((metric) => getBarShape(metric, barArgs)),
          )
        : [],
      source: timeseries,
    }));

  return { shapeLists, xAxis, yAxis };
}

type BarArgs = {
  barWidth: number;
  buckets: StackedChartBuckets;
  isPercentage: boolean;
  xAxis: Axis;
  yAxis: Axis;
};

function getBarShape(
  metric: Metric,
  { xAxis, yAxis, barWidth, isPercentage, buckets }: BarArgs,
): Shape<Metric> | null {
  const bucketValue = buckets.get(metric.time);
  if (!bucketValue || Number.isNaN(metric.value)) {
    return null;
  }

  const time = getTimeFromTimestamp(metric.time);
  const value = isPercentage ? metric.value / bucketValue.total : metric.value;

  const x = normalizeAlongLinearAxis(time, xAxis) - 0.5 * barWidth;
  const y = bucketValue.currentY;

  const height = normalizeAlongLinearAxis(value, yAxis);
  bucketValue.currentY += height;

  return { type: "rectangle", x, y, width: barWidth, height, source: metric };
}
