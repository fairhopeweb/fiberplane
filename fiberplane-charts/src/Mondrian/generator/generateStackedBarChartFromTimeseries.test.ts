import type { Metric, Timeseries } from "../../providerTypes";
import { generateStackedBarChartFromTimeseries } from "./generateStackedBarChartFromTimeseries";

test("it generates a stacked bar chart", () => {
  const timeseriesA: Timeseries = {
    name: "dummy_data",
    labels: { timeseries: "A" },
    attributes: {},
    resource: {},
    metrics: [
      getMetricAtMinute(0, 10),
      getMetricAtMinute(1, 15),
      getMetricAtMinute(2, 20),
      getMetricAtMinute(4, 25),
      getMetricAtMinute(5, 30),
      getMetricAtMinute(6, 25),
      getMetricAtMinute(7, 20),
      getMetricAtMinute(8, 15),
      getMetricAtMinute(9, 10),
    ],
    visible: true,
  };

  const timeseriesB: Timeseries = {
    name: "dummy_data",
    labels: { timeseries: "B" },
    attributes: {},
    resource: {},
    metrics: [
      getMetricAtMinute(0, 40),
      getMetricAtMinute(1, 30),
      getMetricAtMinute(2, 20),
      getMetricAtMinute(3, 10),
      getMetricAtMinute(4, 20),
      getMetricAtMinute(5, 30),
      getMetricAtMinute(6, 40),
      getMetricAtMinute(7, 50),
      getMetricAtMinute(8, 60),
      getMetricAtMinute(9, 70),
    ],
    visible: true,
  };

  expect(
    generateStackedBarChartFromTimeseries({
      graphType: "line",
      stackingType: "stacked",
      timeRange: {
        from: getDateAtMinute(0).toISOString(),
        to: getDateAtMinute(10).toISOString(),
      },
      timeseriesData: [timeseriesA, timeseriesB],
      additionalValues: [40],
    }),
  ).toMatchSnapshot();
});

function getMetricAtMinute(minute: number, value: number): Metric {
  return {
    time: getDateAtMinute(minute).toISOString(),
    value,
    attributes: {},
    resource: {},
  };
}

function getDateAtMinute(minute: number): Date {
  return new Date(
    `2023-07-18T16:${minute.toString().padStart(2, "0")}:00.000Z`,
  );
}
