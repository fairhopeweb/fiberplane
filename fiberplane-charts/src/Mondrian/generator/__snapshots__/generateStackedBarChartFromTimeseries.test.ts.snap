// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it generates a stacked bar chart 1`] = `
{
  "shapeLists": [
    {
      "shapes": [
        {
          "height": 0.125,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:00:00.000Z",
            "value": 10,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.010732323232323232,
          "y": 0,
        },
        {
          "height": 0.1875,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:01:00.000Z",
            "value": 15,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.10164141414141413,
          "y": 0,
        },
        {
          "height": 0.25,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:02:00.000Z",
            "value": 20,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.19255050505050503,
          "y": 0,
        },
        {
          "height": 0.3125,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:04:00.000Z",
            "value": 25,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.3743686868686869,
          "y": 0,
        },
        {
          "height": 0.375,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:05:00.000Z",
            "value": 30,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.4652777777777778,
          "y": 0,
        },
        {
          "height": 0.3125,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:06:00.000Z",
            "value": 25,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.5561868686868687,
          "y": 0,
        },
        {
          "height": 0.25,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:07:00.000Z",
            "value": 20,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.6470959595959596,
          "y": 0,
        },
        {
          "height": 0.1875,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:08:00.000Z",
            "value": 15,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.7380050505050505,
          "y": 0,
        },
        {
          "height": 0.125,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:09:00.000Z",
            "value": 10,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.8289141414141414,
          "y": 0,
        },
      ],
      "source": {
        "attributes": {},
        "labels": {
          "timeseries": "A",
        },
        "metrics": [
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:00:00.000Z",
            "value": 10,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:01:00.000Z",
            "value": 15,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:02:00.000Z",
            "value": 20,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:04:00.000Z",
            "value": 25,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:05:00.000Z",
            "value": 30,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:06:00.000Z",
            "value": 25,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:07:00.000Z",
            "value": 20,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:08:00.000Z",
            "value": 15,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:09:00.000Z",
            "value": 10,
          },
        ],
        "name": "dummy_data",
        "resource": {},
        "visible": true,
      },
    },
    {
      "shapes": [
        {
          "height": 0.5,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:00:00.000Z",
            "value": 40,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.010732323232323232,
          "y": 0.125,
        },
        {
          "height": 0.375,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:01:00.000Z",
            "value": 30,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.10164141414141413,
          "y": 0.1875,
        },
        {
          "height": 0.25,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:02:00.000Z",
            "value": 20,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.19255050505050503,
          "y": 0.25,
        },
        {
          "height": 0.125,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:03:00.000Z",
            "value": 10,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.28345959595959597,
          "y": 0,
        },
        {
          "height": 0.25,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:04:00.000Z",
            "value": 20,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.3743686868686869,
          "y": 0.3125,
        },
        {
          "height": 0.375,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:05:00.000Z",
            "value": 30,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.4652777777777778,
          "y": 0.375,
        },
        {
          "height": 0.5,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:06:00.000Z",
            "value": 40,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.5561868686868687,
          "y": 0.3125,
        },
        {
          "height": 0.625,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:07:00.000Z",
            "value": 50,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.6470959595959596,
          "y": 0.25,
        },
        {
          "height": 0.75,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:08:00.000Z",
            "value": 60,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.7380050505050505,
          "y": 0.1875,
        },
        {
          "height": 0.875,
          "source": {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:09:00.000Z",
            "value": 70,
          },
          "type": "rectangle",
          "width": 0.06944444444444445,
          "x": 0.8289141414141414,
          "y": 0.125,
        },
      ],
      "source": {
        "attributes": {},
        "labels": {
          "timeseries": "B",
        },
        "metrics": [
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:00:00.000Z",
            "value": 40,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:01:00.000Z",
            "value": 30,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:02:00.000Z",
            "value": 20,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:03:00.000Z",
            "value": 10,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:04:00.000Z",
            "value": 20,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:05:00.000Z",
            "value": 30,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:06:00.000Z",
            "value": 40,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:07:00.000Z",
            "value": 50,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:08:00.000Z",
            "value": 60,
          },
          {
            "attributes": {},
            "resource": {},
            "time": "2023-07-18T16:09:00.000Z",
            "value": 70,
          },
        ],
        "name": "dummy_data",
        "resource": {},
        "visible": true,
      },
    },
  ],
  "xAxis": {
    "maxValue": 1689696630,
    "minValue": 1689695970,
    "tickSuggestions": [
      1689696000,
      1689696060,
      1689696120,
      1689696180,
      1689696240,
      1689696300,
      1689696360,
      1689696420,
      1689696480,
      1689696540,
      1689696600,
    ],
  },
  "yAxis": {
    "maxValue": 80,
    "minValue": 0,
  },
}
`;
